activerecord/lib/active_record/tasks/database_tasks.rb

```diff
@@ -409,14 +409,8 @@ def dump_filename(namespace, format = ActiveRecord::Base.schema_format)
         ENV["SCHEMA"] || File.join(ActiveRecord::Tasks::DatabaseTasks.db_dir, filename)
       end
 
-      def cache_dump_filename(namespace)
-        filename = if namespace == "primary"
-          "schema_cache.yml"
-        else
-          "#{namespace}_schema_cache.yml"
-        end
-
-        ENV["SCHEMA_CACHE"] || File.join(ActiveRecord::Tasks::DatabaseTasks.db_dir, filename)
+      def cache_dump_filename(namespace, schema_cache_path = nil)
+        schema_cache_path || ENV["SCHEMA_CACHE"] || default_schema_cache_path(namespace)
```

Comment:
Most of the other methods in `HashConfig` contain their own default values. We could do that here by having the method in the config default to a value based on `spec_name`. Then this line would be something like:

`ENV["SCHEMA_CACHE"] || schema_cache_path`

and we could skip passing `namespace / spec_name` into this method. Alternatively, we might just pass `db_config` into this method and take the value we need of it. Just an idea!

Reply:
I like the idea of having the default in the database config. That's much cleaner!

The db config should still not know about ENV, though, right?

Reply:
Actually, I changed my mind :(

In order to come up with the derived value we need to know the default database directory path, which we would have to get from `Rails.application.config.paths["db"].first` or some such thing. I don't think the database config should reference the top-level Rails config.

Reply:
> Alternatively, we might just pass db_config into this method and take the value we need of it.

I'd prefer not to, since that passes a lot more into the method than it needs.

Overall I'm a bit uncomfortable with the fact that this method needs things from so many places in order to do its job, so it doesn't really _belong_ anywhere!

- `Rails.application.config.paths["db"].first`
- `db_config.spec_name`
- `db_config.schema_cache_path`
- `ENV["SCHEMA_CACHE"]

That's a lot of knowledge.